Zend/zend_compile.c

```diff
@@ -2987,7 +3228,26 @@ void zend_compile_new(znode *result, zend_ast *ast) /* {{{ */
 		class_node.op_type = IS_CONST;
 		ZVAL_STR(&class_node.u.constant, zend_resolve_class_name_ast(class_ast));
 	} else {
-		zend_compile_class_ref(&class_node, class_ast);
+		if (class_ast->kind == ZEND_AST_CLASS) {
+		zend_class_entry *ce = 
+		    zend_compile_class_decl(class_ast TSRMLS_CC);
+		zend_string *name = ce->name;
+		uint32_t fetch_type = zend_get_class_fetch_type(name);
+		
+		opline = zend_emit_op(&class_node,
+		    ZEND_FETCH_CLASS, NULL, NULL TSRMLS_CC);
+			opline->extended_value = fetch_type;
+		
+			if (fetch_type == ZEND_FETCH_CLASS_DEFAULT) {
+			    opline->op2_type = IS_CONST;
+			    opline->op2.constant = zend_add_class_name_literal(CG(active_op_array),
+				zend_resolve_class_name(name, ZEND_NAME_FQ TSRMLS_CC) TSRMLS_CC);
+			}
```

Comment:
For anon classes the fetch_type must always be `ZEND_FETCH_CLASS_DEFAULT` - so this should not be emitting a class fetch and instead directly stick the name in `class_node` (like in the const_default_class_ref case). You also don't need to resolve the name, it's already the canonical one.

There are a few lines above this where the indent is a bit off.